benchmarks/helper.py

```diff
@@ -0,0 +1,241 @@
+                            f"Quick not implemented for {suite}.py"
+                        )
+                lines.append(cmd)
+            lines.append("")
+        runfile.writelines([line + "\n" for line in lines])
+
+
+def pp_dataframe(df, title, output_dir):
+    # Pretty print
+    print(tabulate(df, headers="keys", tablefmt="pretty", showindex="never"))
+
+    # Save to csv, can be copy pasted in google sheets
+    df.to_csv(f"{output_dir}/{title}.csv", index=False)
+
+    # Graph
+    labels = df.columns.values.tolist()
+    labels = labels[2:]
+    df.plot(
+        x="name",
+        y=labels,
+        kind="bar",
+        title=title,
+        ylabel="Speedeup over eager",
+        xlabel="",
+        grid=True,
+        figsize=(max(len(df.index) / 4, 5), 10),
+        edgecolor="black",
+    )
+    plt.tight_layout()
+    plt.savefig(f"{output_dir}/{title}.png")
+
+
+def parse_logs(args, dtypes, suites, devices, compilers, output_dir):
+    mode = "inference" if args.inference else "training"
+    for iter in itertools.product(suites, devices, dtypes):
+        suite, device, dtype = iter
+        frames = []
+        best_compiler = compilers[-1]
+        # Collect results from all the files
+        for compiler in compilers:
+            output_filename = (
+                f"{output_dir}/{compiler}_{suite}_{dtype}_{mode}_{device}.csv"
+            )
+            df = pd.read_csv(output_filename)
+            df.rename(
+                columns={"speedup": compiler, "ts": compiler, "ofi": f"ofi_{compiler}"},
+                inplace=True,
+            )
+            frames.append(df)
+
+        # Merge the results
+        df = pd.merge(*frames, on=["dev", "name"])
+
+        # Pretty print and also write to a bargraph
+        title = f"{suite}_{dtype}_{mode}_{device}"
+        pp_dataframe(df, title, output_dir)
+
+        # Sort the dataframe and pretty print
+        sorted_df = df.sort_values(by=best_compiler, ascending=False)
+        pp_dataframe(sorted_df, f"sorted_{title}", output_dir)
+
+
+if __name__ == "__main__":
+    args = parse_args()
+
+    def extract(key):
+        return DEFAULTS[key] if getattr(args, key, None) is None else getattr(args, key)
+
+    dtypes = extract("dtypes")
+    suites = extract("suites")
+    devices = extract("devices")
+
+    if args.inference:
+        compilers = DEFAULTS["inference"] if args.compilers is None else args.compilers
+    else:  # args.training
+        compilers = DEFAULTS["training"] if args.compilers is None else args.compilers
+
+    output_dir = args.output_dir if args.output_dir is not None else DEFAULT_OUTPUT_DIR
+
+    if args.print_run_commands:
+        generate_commands(args, dtypes, suites, devices, compilers, output_dir)
+    elif args.visualize_logs:
+        parse_logs(args, dtypes, suites, devices, compilers, output_dir)
+    elif args.run:
+        generate_commands(args, dtypes, suites, devices, compilers, output_dir)
+        os.system("bash run.sh")
```

Comment:
does this do enough to flag / catch if run.sh errs?

Reply:
Ah, missed that. Thanks!